design/authenticators/application-identity/application-identity.md

```diff
@@ -0,0 +1,464 @@
+
+In conclusion, our general ValidateApplication class will need the following
+inputs:
+  - Authenticator name (e.g 'authn-azure')
+  - Authenticator's service-id (e.g 'prod')
+  - host's annotations
+  - Different lists of annotation names
+    - permitted
+    - required
+    - non-permitted combinations
+  - Authentication request Application Identity object
+  
+It will perform the following steps:
+  1. Build the Host Application Identity object using the:
+     - authenticator name
+     - authenticator service-id
+     - host's annotations
+     - list of permitted annotations
+  2. Verify the Host Application Identity is configured correctly using the:
+     - Host Application Identity object built in the previous step
+     - list of required constraints
+     - list of non-permitted constraints combinations
+  3. Verify the Host Application Identity matches the authentication request
+   using the:
+     - Host Application Identity object built in the first step
+     - Authentication request Application Identity object given in the input
+      to this class
+      
+Let's see how the classes will look, in high level.
+
+#### Application Identity class
+
+We will have 2 different Application Identity classes. Although their
+interface is the same, they are not entirely identical. 
+
+At first, the `constraints` method will return a simple key-value hash. We
+should challenge that and find a better way to do it in Ruby.
+
+##### Host Application Identity
+
+```
+class HostApplicationIdentity
+
+      def initialize(authenticator_name:, service_id:, role_annotations:, permitted_annotations:)
+        @authenticator_name = authenticator_name
+        @service_id       = service_id
+        @role_annotations = role_annotations
+        @permitted_annotations = permitted_annotations
+      end
+
+      def constraints
+        @constraints ||= {}
+
+        @permitted_annotations.each { |annotation_name|
+          @constraints[annotation_name] = constraint_value(annotation_name)
+        }
+      end
+
+      private
+
+      # check the `service-id` specific constraint first to be more granular
+      def constraint_value constraint_name
+        annotation_value("#{@authenticator_name}/#{@service_id}/#{constraint_name}") ||
+          annotation_value("#{@authenticator_name}/#{constraint_name}")
+      end
+
+      def annotation_value name
+        annotation = @role_annotations.find { |a| a.values[:name] == name }
+
+        # return the value of the annotation if it exists, nil otherwise
+        if annotation
+          @logger.debug(LogMessages::Authentication::RetrievedAnnotationValue.new(name))
+          annotation[:value]
+        end
+      end
+    end
+```
+
+##### Authentication request Application Identity
+
+The input for this class will be built in each authenticator as it differs
+ between them. For simplicity, the class will receive the `constraints` hash
+  as an input. We should challenge that to find a better way.
```

Comment:
what other options do we have? a separate object?

Reply:
yes. i'll add that.

Reply:
oh no, i thought you meant something else. I actually don't have yet a better idea but it seems that we can find one if we look for it.